### PR TITLE
bugfix - Eagerly load cache on running in simulate/spy mode

### PR DESCRIPTION
**Pull request by kapishmalik**
Resolves => #1002 
@tommysitu  could you please review.

bugfix - Eagerly load cache on running in simulate/spy mode using hoverfly cmd

### CODE REVIEW

**Comment by tommysitu** [comment]
Hi @kapishmalik thank you for looking into this bug. Could I suggest a few improvements please?

1. It'd be useful to see if the cache has been preloaded or not right from the command line, so logging an info would be really really helpful. 
2.  We only preload cache if the simulation has exact matchers only, but since v1.5.0, capturing query params could result in an array matcher, so the preloading won't work any more if that's the case. It'd be good to get it fixed too.
3. When using `hoverctl`, only calling `hoverctl mode simulate` will preload, but not using `hoverctl import`. In order to bring hoverctl to parity with hoverfly commands, we should fix `hoverctl import` as well
4. Would be good to add some functional tests for cache pre-loading to prevent future regression.

I hope it's not too much to ask for!

**Comment by kapishmalik** [comment]
Sure @tommysitu, I will incorporate all of these and get back.

**Comment by tommysitu** [comment]
> Sure @tommysitu, I will incorporate all of these and get back.

You're star

File: core/matching/cache_matcher.go [hunk 1]
```diff
@@ -150,5 +151,6 @@ func (this *CacheMatcher) PreloadCache(simulation *models.Simulation) error {
 		}
 	}
 
+	log.Info("Exiting after preloading cache")
```
Comment (tommysitu) [inline core/matching/cache_matcher.go:154]: Something like `Cache preloaded` is ok

File: core/matching/cache_matcher.go [hunk 2]
```diff
@@ -139,6 +139,7 @@ func (this *CacheMatcher) FlushCache() error {
 }
 
 func (this *CacheMatcher) PreloadCache(simulation *models.Simulation) error {
+	log.Info("Entered for preloading cache")
```
Comment (tommysitu) [inline core/matching/cache_matcher.go:142]: not needed

Comment (kapishmalik) [reply core/matching/cache_matcher.go]: done

**Comment by kapishmalik** [comment]
I need to test and add/fix UTs and FTs. I will inform here once done.

**Comment by kapishmalik** [comment]
@tommysitu I started writing FTs but found one FT saying that the cache should be flushed when importing via API. So, should we incorporate point 3 i.e. to preload the cache on doing the hoverctl import command that eventually call PUT api. 

```It("should be flushed when importing via the API", func() {
		hoverfly.ImportSimulation(testdata.JsonPayload)
		cacheView := hoverfly.GetCache()

		Expect(cacheView.Cache).To(HaveLen(0))
	})

**Comment by kapishmalik** [comment]
@tommysitu it ll be great if you can help me with all scenarios when cache should be preloaded? I will document also so that expectations should be clear from hoverfly tool of the users.

**Comment by tommysitu** [comment]
right, maybe I've missed something, let's skip `3` for now. Thanks!

**Comment by kapishmalik** [comment]
@tommysitu I have incorporated all the comments/suggestions provided above. Could you please check this PR.

File: core/matching/cache_matcher.go [hunk 3]
```diff
@@ -150,5 +151,6 @@ func (this *CacheMatcher) PreloadCache(simulation *models.Simulation) error {
 		}
 	}
 
+	log.Info("Cache preloaded")
```
Comment (tommysitu) [inline core/matching/cache_matcher.go:154]: this log should go into the `if` block where the cache is actually pre-loaded

File: core/models/request_matcher.go [hunk 4]
```diff
@@ -313,10 +313,15 @@ func (this RequestMatcher) ToEagerlyCacheable() *RequestDetails {
 	if this.Query != nil && len(*this.Query) > 0 {
 		for key, valueMatchers := range *this.Query {
 			for _, valueMatcher := range valueMatchers {
-				if valueMatcher.Matcher != matchers.Exact {
+				if valueMatcher.Matcher != matchers.Exact && !(valueMatcher.Matcher == matchers.Array && (valueMatcher.Config == nil || len(valueMatcher.Config) == 0)) {
 					return nil
 				}
-				query[key] = []string{valueMatcher.Value.(string)}
+				if valueMatcher.Matcher == matchers.Array {
+					query[key] = valueMatcher.Value.([]string)
```
Comment (tommysitu) [inline core/models/request_matcher.go:320]: this line throws an array when I import a simulation with query array matcher: 
```
panic: interface conversion: interface {} is []interface {}, not []string

goroutine 1 [running]:
github.com/SpectoLabs/hoverfly/core/models.RequestMatcher.ToEagerlyCacheable({{0x140003e4810, 0x1, 0x1}, {0x140003e4840, 0x1, 0x1}, {0x140003e4870, 0x1, 0x1}, {0x140003e48a0, ...}, ...})
        /Users/tommy.situ/repos/kapish/hoverfly/core/models/request_matcher.go:320 +0x958
github.com/SpectoLabs/hoverfly/core/matching.(*CacheMatcher).PreloadCache(0x140002efd60, 0x16dd5b5f2?)
        /Users/tommy.situ/repos/kapish/hoverfly/core/matching/cache_matcher.go:148 +0xdc
main.main()
        /Users/tommy.situ/repos/kapish/hoverfly/core/cmd/hoverfly/main.go:554 +0x2ee4

```

You can try capture a simulation using this:
`curl --proxy localhost:8500 'time.jsontest.com?token=1&token=2'`

File: core/matching/cache_matcher.go (hunk 3 above)
Comment (kapishmalik) [reply core/matching/cache_matcher.go:154]: got it... done

File: core/models/request_matcher.go (hunk 4 above)
Comment (kapishmalik) [reply core/models/request_matcher.go]: incorporated

**Review by tommysitu — APPROVED** [review]
noice!

**Pull request merged**